Support for X25519Kyber768Draft00 post-quantum key exchange

ctz opened the pull request:
Draft for the moment.  This is on top of #1784.

I probably won't do `rustls-postquantum` as a separate crate, but it made sense at the time.

Most likely this will end up inside rustls/crypto/aws_lc_rs/ somewhere. I don't think it should be used by default for now, given it is pre-standardisation.

djc commented on rustls-postquantum/Cargo.toml:
```diff
@@ -7,3 +7,7 @@ publish = false
 aws-lc-rs = { version = "1.6", features = [ "unstable" ], default-features = false }
+
+[dev-dependencies]
+env_logger = "0.10"
```
Add the comment caveat about avoiding the upgrade?

djc reviewed (commented):
Cool!

cpu commented on rustls/src/crypto/mod.rs:
```diff
@@ -371,6 +371,28 @@ pub trait SupportedKxGroup: Send + Sync + Debug {
+    /// function should be implemented.
+    fn start_and_complete(&self, peer_pub_key: &[u8]) -> Result<CompletedKeyExchange, Error> {
+        let kx = self.start()?;
+        let group = kx.group();
+        let pub_key = kx.pub_key().to_vec();
+        let secret = kx.complete(peer_pub_key)?;
+
+        Ok(CompletedKeyExchange {
+            group,
+            pub_key,
+            secret,
+        })
```
nit: since you have to create all of these bindings just to use them for the struct field initialization it feels more slightly more natural to do it in-line.
```suggestion
        Ok(CompletedKeyExchange {
            group: kx.group(),
            pub_key: kx.pub_key().to_vec(),
            secret: kx.complete(peer_pub_key)?,
        })
```

cpu commented on rustls/src/crypto/mod.rs:
```diff
@@ -371,6 +371,28 @@ pub trait SupportedKxGroup: Send + Sync + Debug {
+    ///
+    /// If there is such a data dependency (like key encapsulation mechanisms), this
+    /// function should be implemented.
+    fn start_and_complete(&self, peer_pub_key: &[u8]) -> Result<CompletedKeyExchange, Error> {
```
Revisiting [semver guidance](https://doc.rust-lang.org/cargo/reference/semver.html#trait-new-default-item) this is "Possibly-breaking", but only if the same method exists in another trait that would be implemented on a type implementing `CryptoProvider`, and that feels unlikely :+1:

cpu commented on rustls/src/crypto/mod.rs:
```diff
@@ -455,6 +477,18 @@ pub trait ActiveKeyExchange: Send + Sync {
 }
 
+/// The result from [`SupportedKxGroup::start_and_complete()`].
+pub struct CompletedKeyExchange {
```
I wondered if this should have a `Debug` impl, but it would need to be custom: `SharedSecret` doesn't impl `Debug` (and shouldn't) - so probably not worth it.

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -358,10 +358,10 @@ pub struct KeyShareEntry {
 
 impl KeyShareEntry {
-    pub fn new(group: NamedGroup, payload: &[u8]) -> Self {
+    pub fn new(group: NamedGroup, payload: impl Into<Vec<u8>>) -> Self {
```
I think we can ignore semver compat here because `KeyShareEntry` is only exposed via the `internal` namespace.

cpu commented on rustls-postquantum/src/lib.rs:
```diff
@@ -1,13 +1,132 @@
     }
 }
+
+///
```
TODO?

ctz replied:
added docs for this

cpu commented on rustls-postquantum/examples/client.rs:
```diff
@@ -0,0 +1,51 @@
+fn main() {
+    env_logger::init();
+
+    let mut root_store = rustls::RootCertStore::empty();
+    root_store.extend(
+        webpki_roots::TLS_SERVER_ROOTS
+            .iter()
+            .cloned(),
+    );
```
nit:
```suggestion
    let root_store = RootCertStore {
        roots: webpki_roots::TLS_SERVER_ROOTS.into(),
    };
```

ctz replied:
also done this to simpleclient.rs

cpu commented on rustls-postquantum/examples/client.rs:
```diff
@@ -0,0 +1,51 @@
+use std::io::{stdout, Read, Write};
```
Can you add a short rustdoc comment describing the purpose of the example here?

ctz replied:
added a comment about the purpose and what to look for in its output

xuganyu96 commented:
Hi all, I am working on a project experimenting with integrating post-quantum KEMs into `rustls`, and this PR's proposed addition of `start_and_complete` to `ActiveKeyExchange` seems to really help. Would it be possible to merge a subset of this PR, including `ActiveKeyExchange::start_and_complete` and the changes to `emit_server_hello`?

cpu commented:
>  Would it be possible to merge a subset of this PR

I think it would probably be easier for you to achieve your goal by taking a Cargo patch on this branch or a specific rev in it, no? You'll have to do the same thing to pick up changes in `main` that aren't in a released version.

ctz commented:
Have pushed a new revision of this:

- addressing the above review comments
- added quite a bit of commentary at the crate level
- made explicit ciphertext/share types in instead of lots of slicing

I think I like the idea of releasing this as a separate crate for now, because:

- it's pre-standardisation, so might have a short lifetime and different release cadence.
- it implies enabling the `unstable` feature on `aws-lc-rs`, I don't want to do this for the main crate (or add a new feature).

Once those two become untrue, I think we can migrate it into the main crate and sunset `rustls-postquantum`.

cpu commented on rustls-postquantum/src/lib.rs:
```diff
@@ -0,0 +1,231 @@
+//!
+//! X25519Kyber768Draft00 is pre-standardization, so you should treat
+//! this as experimental.  You may see unexpected interop failures, and
+//! the algorithm implemented here may not be the one eventually
+//! becomes widely deployed.
```
typo: missing a word here I think:

```suggestion
//! the algorithm implemented here may not be the one that eventually
//! becomes widely deployed.
```

cpu commented on rustls-postquantum/src/lib.rs:
```diff
@@ -0,0 +1,231 @@
+//!
+//! However, the two components of this key exchange are well regarded:
+//! X25519 alone is already used by default by rustls, and tends to have
+//! higher quality implementations that other elliptic curves.
```
typo:

```suggestion
//! higher quality implementations than other elliptic curves.
```

cpu commented on rustls-postquantum/src/lib.rs:
```diff
@@ -0,0 +1,231 @@
+    }
+}
+
+/// This is [X25519Kyber768Draft00] key exchange.
```
nit: feels like this reads better with "the" but I don't feel strongly:

```suggestion
/// This is the [X25519Kyber768Draft00] key exchange.
```

cpu commented on rustls-postquantum/src/lib.rs:
```diff
@@ -0,0 +1,231 @@
+    }
+}
+
+const NAMED_GROUP: NamedGroup = NamedGroup::Unknown(0x6399);
```
I think it's unrelated to this branch, but I wonder if the `Unknown` variant could have a more descriptive name. In this context it _is_ known, just not to Rustls, but to the extension crypto provider :thinking: 

Maybe `NamedGroup::Extension`? I'll give it more thought. Just rubber ducking here.

ctz replied:
`Other` perhaps?

cpu replied:
I like it, I'll put up a separate rename PR so we can resolve this thread.

ctz replied:
I think this would be a breaking change -- maybe we can leave it for later?

ctz replied:
(Or alternatively we could branch for 0.23 maintenance.)

cpu commented on rustls-postquantum/examples/client.rs:
```diff
@@ -0,0 +1,58 @@
+//! This is the simplest possible client using rustls-postquantum, based on
```
Would you mind also adding a `cargo run --locked --bin rustls-postquantum` to the `.github/workflows/daily-tests.yml`'s `example-tests` section? It could go alongside the `simpleclient`, `limitedclient`, etc steps.

ctz replied:
Good point, have added this, with a `grep` at the end to check this basically works.

ctz replied:
Hah, that was surprisingly annoying. It seems some cloudflare POPs serve different pages for pq.cloudflareresearch.com - in the UK I get a static HTML page, but CI in the US gets javascript 🤨 

So now this hits the API endpoint that the javascript targets.

cpu replied:
Tricksy CDNs and their tricksy games :roll_eyes:

djc commented:
The preferred spelling for post-quantum appears to use a hyphen, so I'd expect the crate to be named rustls-post-quantum instead?

cpu reviewed (commented):
Nice! I just had some small nits to flag.

djc commented on rustls-postquantum/src/lib.rs:
```diff
@@ -1,14 +1,231 @@
+    }
+}
+
+fn kyber768_r3() -> &'static kem::Algorithm<AlgorithmId> {
```
Nit: move this down, to just above the consts?

djc commented on rustls-postquantum/Cargo.toml:
```diff
@@ -0,0 +1,8 @@
+publish = false
+
+[dependencies]
+rustls = { path = "../rustls", features = [ "aws_lc_rs" ]}
```
Nit: I think convention is not to have spaces inside `[]` for features (also in later commits).

djc reviewed (approved):
Cool!

Some nits, plus the thing I mentioned about including a dash in the new crate's name.

ctz commented:
> so I'd expect the crate to be named rustls-post-quantum instead?

Renamed.
